feedback + card.uuid - first take (#519)
cds-hooks/docs: docs/specification/current.md

Field | Optionality | Type | Description
----- | ----- | ----- | --------

> `uuid` | OPTIONAL | *string* | Unique identifier of the card, used for auditing and logging cards and to provide feedback.

Reviewer: It's unclear to me how a uuid provides feedback, consider striking 'and to provide feedback'

Author: Brett, the newly added `card.uuid` is primarily needed for the cds client to identify which card it's talking about when it calls the cds service's feedback endpoint. I've tried to clarify that because, to your point, a uuid doesn't provide feedback. :)

If you still think 'and to provide feedback should be scratched, let me know and I'll do so.

Author: Brett,

The primary purpose of card.uuid is to identify which card the cds client is talking about when it subsequently calls a cds service's feedback endpoint. I attempted to clarify this, let me know if you still think this phrase should be removed.

Isaac

> `uuid` | OPTIONAL | *string* | Unique identifier of the card, used for auditing and logging cards and to provide feedback.

Author: ```suggestion
`uuid` | OPTIONAL | *string* | Unique identifier of the card, used for auditing and logging cards and included in any subsequent calls to the CDS service's feedback endpoint.
```

> `uuid` | OPTIONAL | *string* | Unique identifier of the card, used for auditing and logging cards and included in any subsequent calls to the CDS service's feedback endpoint.

Reviewer 2: ```suggestion
`uuid` | OPTIONAL | *string* | Unique identifier of the card.  MAY be used for auditing and logging cards and SHALL be included in any subsequent calls to the CDS service's feedback endpoint.
```

}
```

## Feedback

> Once a CDS Hooks service responds to a hook by returning an info or suggestion card, the service has no further interaction with the CDS client. The acceptance of a suggestion or rejection of a card is valuable information to enable a service to improve its behavior towards the goal of the end-user having a positive and meaningful experience with the CDS. A feedback endpoint enables suggestion tracking & analytics.

Reviewer 2: ```suggestion
Once a CDS Hooks service responds to a hook by returning a card, the service has no further interaction with the CDS client. The acceptance of a suggestion or rejection of a card is valuable information to enable a service to improve its behavior towards the goal of the end-user having a positive and meaningful experience with the CDS. A feedback endpoint enables suggestion tracking & analytics.
```

> Upon receiving a card, a user may accept its sugestions, ignore it entirely, or dismiss it with or without an override reason. Note that while one or more suggestions can be accepted, an entire card is either ignored or overridden.

Reviewer 2: ```suggestion
Upon receiving a card, a user may accept its suggestions, ignore it entirely, or dismiss it with or without an override reason. Note that while one or more suggestions can be accepted, an entire card is either ignored or overridden.
```

Typically, an end user may only accept (a suggestion), or override a card once; however, a card once ignored could later be acted upon. CDS Hooks does not specify the UI behavior of CDS clients, including the persistence of cards. CDS clients should faithfully report each of these distinct end-user interactions as feedback. 

Each **Feedback** is described by the following attributes.

Field | Optionality | Type | Description
----- | ----- | ----- | --------
`card` | REQUIRED | *string* | The `card.uuid` from the CDS Hooks response. Uniquely identifies the card.
`outcome` | REQUIRED | *string* | A value of `accepted` or `overridden`.

> `acceptedSuggestions` | CONDITIONAL | *array* | An array of json objects identifying one or more suggestions accepted by the user. Required for `accepted` outcomes.

Reviewer 2: While we have a table for OverrideReasons, we don’t have anything for an accepted suggestion.

Author: Yeah, I considered doing this, but there's (currently) only a single attribute in the acceptedSuggestions object. 

A table with a single row seems of less value, but I'll add it anyway.

> `overrideReason` | OPTIONAL | *array* | A json object identifying the `overrideReason` selected by the user.

Reviewer 2: ```suggestion
`overrideReason` | OPTIONAL | *array* | A json object identifying the **OverrideReason** selected by the user.
```

> `overrideReason` | OPTIONAL | *array* | A json object identifying the `overrideReason` selected by the user.

Reviewer 2: ```suggestion
`overrideReason` | OPTIONAL | *object* | A json object identifying the `overrideReason` selected by the user.
```

Author: in the example, overrideReasons is an array of objects. Do you still think this should be an object data type?

Author: Aw, you're right. Sorry about that.

`outcomeTimestamp` | REQUIRED | *string* | ISO timestamp in UTC when action was taken on card.

## Suggestion accepted

The CDS client can inform the service when one or more suggestions were accepted by POSTing a simple json object. The CDS client authenticates to the CDS service as described in [Trusting CDS Clients](#trusting-cds-clients).

Upon the user accepting a suggestion (perhaps when she clicks a displayed label (e.g., button) from a "suggestion" card), the CDS client informs the service by posting the card and suggestion `uuid`s to the CDS Service's feedback endpoint with an outcome of `accepted`.

> To enable a positive clinical experience, the analytics endpoint may be called for multiple hook instances or multiple cards at the same time or even multiple times for a card or suggestion. Depending upon the UI and workflow of the CDS client, a CDS Service may receive feedback for the same card instance multiple times. 

Reviewer 2: ```suggestion
To enable a positive clinical experience, the feedback endpoint may be called for multiple hook instances or multiple cards at the same time or even multiple times for a card or suggestion. Depending upon the UI and workflow of the CDS client, a CDS Service may receive feedback for the same card instance multiple times. 
```

Author: great catch, Dennis!

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[
      {
         "card":"f6b95768-b1c8-40dc-8385-bf3504b82ffb", // uuid from `card.uuid`
         "outcome":"overridden",
         "outcomeTimestamp": "iso timestamp in UTC when action was taken on card"
      }
   ]
}

```

## Explicit reject with override reasons

A CDS client can inform the service when a card was rejected by POSTing an outcome of `overridden` along with an `overrideReason` to the service's feedback endpoint. The CDS Client may enable the clinician to supplement the `overrideReason` with a free text comment, supplied to the CDS Service in `overrideReason.userComment`. 

Each **OverrideReason** is described by the following attributes, in the feedback POST to the CDS Service.

Field | Optionality | Type | Description
----- | ----- | ----- | --------

> `code` | REQUIRED | *string* | The code for this OverrideReason as provided by the CDS Service in the CDS Hooks response.

Reviewer 2: Make this conditional?  The user may just supply a comment

> `system` | CONDITIONAL | *string* | The codesystem for this OverrideReason code as provided by the CDS Service in the CDS Hooks response. Required if the CDS Service supplied this information in the  CDS Hooks response.

Reviewer 2: ```suggestion
`system` | CONDITIONAL | *string* | The codesystem for this OverrideReason code as provided by the CDS Service in the CDS Hooks response. Required if the user selected an overrideReason and the CDS Service supplied this information in the CDS Hooks response.
```

> `userComment` | OPTIONAL | *array* | The CDS Client may enable the clinician to further explain why the card was rejected with free text. That user comment may be communicated to the CDS Service as a `userComment`.

Reviewer 2: Is this mean to be an array?  The user entering multiple comments?

Reviewer 2: Opened this thread back up

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[{
         "card":"9368d37b-283f-44a0-93ea-547cebab93ed",
         "outcome":"overridden",
         "overrideReason": { 
	 	"code":"reason-code-provided-by-service",
     		"system":"http://example.org/cds-services/fhir/CodeSystem/override-reasons",

> 		"userComment" : "clinician entered comment>" 

Reviewer 2: ```suggestion
		"userComment" : "clinician entered comment" 
```

Upon receiving a card, a user may accept its suggestions, ignore it entirely, or dismiss it with or without an override reason. Note that while one or more suggestions can be accepted, an entire card is either ignored or overridden.

Typically, an end user may only accept (a suggestion), or override a card once; however, a card once ignored could later be acted upon. CDS Hooks does not specify the UI behavior of CDS clients, including the persistence of cards. CDS clients should faithfully report each of these distinct end-user interactions as feedback. 

Each **Feedback** is described by the following attributes.

Field | Optionality | Type | Description
----- | ----- | ----- | --------
`card` | REQUIRED | *string* | The `card.uuid` from the CDS Hooks response. Uniquely identifies the card.
`outcome` | REQUIRED | *string* | A value of `accepted` or `overridden`.

> `acceptedSuggestions` | CONDITIONAL | *array* | An array of json objects identifying one or more suggestions accepted by the user. Required for `accepted` outcomes.

Reviewer 2: ```suggestion
`acceptedSuggestions` | CONDITIONAL | *array* | An array of json objects identifying one or more of the user's **AcceptedSuggestion**s. Required for `accepted` outcomes.
```

Upon the user accepting a suggestion (perhaps when she clicks a displayed label (e.g., button) from a "suggestion" card), the CDS client informs the service by posting the card and suggestion `uuid`s to the CDS Service's feedback endpoint with an outcome of `accepted`.

To enable a positive clinical experience, the analytics endpoint may be called for multiple hook instances or multiple cards at the same time or even multiple times for a card or suggestion. Depending upon the UI and workflow of the CDS client, a CDS Service may receive feedback for the same card instance multiple times. 

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[
      {
         "card":"`card.uuid` from CDS Hooks response",
         "outcome":"accepted",
         "acceptedSuggestions": [ { "id" : "`card.suggestion.uuid` from CDS Hooks response" } ],
         "outcomeTimestamp": "iso timestamp in UTC when action was taken on card"
      }
   ]
}
```

If either the card or the suggestion has no `uuid`, the CDS client does not send a notification.

## Card ignored

> If the end-user doesn't interact with the CDS Service's card at all, the card is *ignored*. Why would this happen? Perhaps the priority indicator of the card deprioritized it, or the user simply ignored the guidance. In this case, the CDS Client does not inform the CDS Service of the rejected guidance. Even with a `card.uuid`, a `suggestion.uuid`, and an available feedback service, the service is not informed. 

Reviewer: strike the question and response "Why would this happen? Perhaps the priority indicator of the card deprioritized it, or the user simply ignored the guidance."

Author: done

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[
      {
         "card":"f6b95768-b1c8-40dc-8385-bf3504b82ffb", // uuid from `card.uuid`
         "outcome":"overridden",
         "outcomeTimestamp": "iso timestamp in UTC when action was taken on card"
      }
   ]
}

```

## Explicit reject with override reasons

A CDS client can inform the service when a suggestion was rejected by POSTing an outcome of `overridden` along with an `overrideReason` to the service's feedback endpoint. The CDS Client may enable the clinician to supplement the `overrideReason` with a free text comment, supplied to the CDS Service in `overrideReason.userComment`. 

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[
      {
         "card":"9368d37b-283f-44a0-93ea-547cebab93ed",
         "outcome":"overridden",

>          "overrideReason": { "key" : "reason-id-provided-by-service", "userComment" : "clinician entered comment>" }

Reviewer 2: 'key' is left over from an earlier version of the overrideReasons PR.  Now we need the system+code

}
```

> ## Feedback

Reviewer 2: I think it'd be helpful, as well as consist with the rest of the spec, to add a table specifying the fields in the request to the feedback service, along with optionality, type, and description

Author: Dennis, this is done. Will you re-review?

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[
      {
         "card":"f6b95768-b1c8-40dc-8385-bf3504b82ffb", // uuid from `card.uuid`
         "outcome":"overridden",
         "outcomeTimestamp": "iso timestamp in UTC when action was taken on card"
      }
   ]
}

```

## Explicit reject with override reasons

A CDS client can inform the service when a suggestion was rejected by POSTing an outcome of `overridden` along with an `overrideReason` to the service's feedback endpoint. The CDS Client may enable the clinician to supplement the `overrideReason` with a free text comment, supplied to the CDS Service in `overrideReason.userComment`. 

```
POST {baseUrl}/cds-services/{serviceId}/feedback

{
   "feedback":[{
         "card":"9368d37b-283f-44a0-93ea-547cebab93ed",
         "outcome":"overridden",

>          "overrideReason": { 

Reviewer 3: Would it be acceptable to send multiple override reasons per card for each call to the `feedback` endpoint?  This current example shows just 1.

Reviewer 2: I think for now we just consider selecting one unless there's a use-case for multiple.
